php-checks/src/main/java/org/sonar/php/checks/UseOfUninitializedVariableCheck.java

```diff
@@ -87,175 +97,239 @@
+        result.add(((VariableIdentifierTree) tree.key()).variableExpression().text());
+      } else {
+        TreeUtils.descendants(tree.key(), VariableIdentifierTree.class).forEach(v -> result.add(v.variableExpression().text()));
       }
-      super.visitFunctionCall(functionCall);
     }
 
-    @Override
-    public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
-      if (tree == currentFunction) {
-        super.visitFunctionDeclaration(tree);
-      }
-      // else skip nested
+    return result;
+  }
+
+  private static BlockSummary getBlockSummary(CfgBlock block, BlockSummary predecessorsSummary) {
+    BlockSummary summary = BlockSummary.copy(predecessorsSummary);
+
+    for (Tree element : block.elements()) {
+      SummaryUpdateVisitor visitor = new SummaryUpdateVisitor(summary);
+      element.accept(visitor);
     }
 
-    @Override
-    public void visitFunctionExpression(FunctionExpressionTree tree) {
-      LexicalVariablesTree lexicalVars = tree.lexicalVars();
-      if (tree == currentFunction) {
-        if (lexicalVars != null) {
-          lexicalVars.variables().stream()
-            .map(VariableTree::variableExpression)
-            .filter(variable -> variable.is(Kind.VARIABLE_IDENTIFIER))
-            .map(variable -> ((VariableIdentifierTree) variable).text())
-            .forEach(initializedVariables::add);
-        }
-        super.visitFunctionExpression(tree);
-      } else {
-        if (lexicalVars != null) {
-          scan(lexicalVars.variables());
-        }
-        // skip nested
+    if (block instanceof CfgBranchingBlock) {
+      Tree branchingTree = ((CfgBranchingBlock) block).branchingTree();
+      if (branchingTree.is(Kind.FOREACH_STATEMENT, Kind.ALTERNATIVE_FOREACH_STATEMENT)) {
+        summary.initializedInBlock.addAll(getForEachVariables((ForEachStatementTree) branchingTree));
       }
     }
 
-    @Override
-    public void visitClassDeclaration(ClassDeclarationTree tree) {
-      // skip nested
-    }
+    return summary;
+  }
 
-    @Override
-    public void visitAnonymousClass(AnonymousClassTree tree) {
-      // skip nested
+  private static class BlockSummary {
+    private final Set<String> initializedFromPredecessors;
+    private final Set<String> initializedInBlock;
+    private boolean scopeWasChanged;
+
+    private BlockSummary(Set<String> initializedFromPredecessors,
+                         Set<String> initializedInBlock,
+                         boolean scopeWasChanged) {
+      this.initializedFromPredecessors = initializedFromPredecessors;
+      this.initializedInBlock = initializedInBlock;
+      this.scopeWasChanged = scopeWasChanged;
     }
 
-    private static boolean isClassMemberAccess(Tree tree) {
-      Tree child = skipParentArrayAccess(tree);
-      return (child.getParent().is(Kind.CLASS_MEMBER_ACCESS) &&
-        ((MemberAccessTree) child.getParent()).member() == child);
+    private static BlockSummary copy(BlockSummary toCopy) {
```

Comment:
My previous comment was about making this method non-static. Sorry if it was not clear.